## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#ab328b6e-06fb-40c7-a058-9a1165efe0e5) |
| mint-xl.sh | :x: [more...](#c7117105-6f92-4cd3-bae8-7296d66cf7ca) |
| mint-large-bucket.sh | :x: [more...](#ffcc0600-c2c2-408e-b444-7af757e2da99) |
| mint-compression-dist-xl.sh | :x: [more...](#cc1195d4-8d1e-4690-a458-a6347bba0b2b) |
<a name="ab328b6e-06fb-40c7-a058-9a1165efe0e5"></a>
### 7950-7434a73/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:30596
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 31a2cf3ee138:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 41 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 59 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 33 seconds
(8/14) Running minio-dotnet tests ... done in 26 seconds
(9/14) Running minio-go tests ... done in 40 seconds
(10/14) Running minio-java tests ... FAILED in 2 seconds
{
  "name": "minio-java",
  "function": "putObject(String bucketName, String objectName, String filename)",
  "args": "filename: 6MB",
  "duration": 1282,
  "status": "FAIL",
  "error": "java.io.EOFException >>> [okio.RealBufferedSink.write(RealBufferedSink.java:115), io.minio.HttpRequestBody.writeTo(HttpRequestBody.java:77), okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:83), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121), okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121), okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:125), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121), okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:264), okhttp3.RealCall.execute(RealCall.java:93), io.minio.MinioClient.executeReq(MinioClient.java:1111), io.minio.MinioClient.execute(MinioClient.java:1070), io.minio.MinioClient.executePut(MinioClient.java:1405), io.minio.MinioClient.executePut(MinioClient.java:1427), io.minio.MinioClient.putObject(MinioClient.java:4216), io.minio.MinioClient.putObject(MinioClient.java:4328), io.minio.MinioClient.putObject(MinioClient.java:3591), FunctionalTest.putObject_test2(FunctionalTest.java:440), FunctionalTest.runTests(FunctionalTest.java:2957), FunctionalTest.main(FunctionalTest.java:3118)]"
}

Executed 9 out of 14 tests successfully.
```
<a name="c7117105-6f92-4cd3-bae8-7296d66cf7ca"></a>
### 7950-7434a73/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:30516
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c75e29559d43:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-dhfluccjeehddt",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 360015,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 15B2DBA1B6808E9A, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="cc1195d4-8d1e-4690-a458-a6347bba0b2b"></a>
### 7950-7434a73/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:30465
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 41c73eca8281:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 3 seconds
(5/14) Running awscli tests ... done in 1 minutes and 4 seconds
(6/14) Running healthcheck tests ... done in 1 seconds
(7/14) Running mc tests ... done in 46 seconds
(8/14) Running minio-dotnet tests ... done in 40 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 8 seconds
(10/14) Running minio-java tests ... done in 58 seconds
(11/14) Running minio-js tests ... done in 37 seconds
(12/14) Running minio-py tests ... FAILED in 17 minutes and 30 seconds
{
  "duration": 103,
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/response.py\", line 397, in _error_catcher\n    yield\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/response.py\", line 475, in read\n    data = self._fp.read()\n  File \"/usr/lib/python3.5/http/client.py\", line 461, in read\n    s = self._safe_read(self.length)\n  File \"/usr/lib/python3.5/http/client.py\", line 609, in _safe_read\n    raise IncompleteRead(b''.join(s), amt)\nhttp.client.IncompleteRead: IncompleteRead(0 bytes read, 1024 more expected)\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 619, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/response.py\", line 563, in from_httplib\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/response.py\", line 238, in __init__\n    self._body = self.read(decode_content=decode_content)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/response.py\", line 496, in read\n    raise IncompleteRead(self._fp_bytes_read, self.length_remaining)\n  File \"/usr/lib/python3.5/contextlib.py\", line 77, in __exit__\n    self.gen.throw(type, value, traceback)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/response.py\", line 415, in _error_catcher\n    raise ProtocolError('Connection broken: %r' % e, e)\nurllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(0 bytes read, 1024 more expected)', IncompleteRead(0 bytes read, 1024 more expected))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1282, in test_presigned_get_object_response_headers\n    response = _http.urlopen('GET', presigned_get_object_url)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/poolmanager.py\", line 326, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 670, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 670, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 670, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 641, in urlopen\n    _stacktrace=sys.exc_info()[2])\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/util/retry.py\", line 399, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='72.28.97.55', port=30465): Max retries exceeded with url: /minio-py-test-a169171a-ffc6-4af4-92ef-8a6aab49a0f4/dca89c2a-69bb-4dbd-a2e7-3045402f4dda?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20190719%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190719T164239Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&response-content-language=en_US&response-content-type=text%2Fplain&X-Amz-Signature=bcc6faff1fe17d1779e4f6f89ac95367a29c7ae25e43ef3b02dbf3593e90b64b (Caused by ProtocolError('Connection broken: IncompleteRead(0 bytes read, 1024 more expected)', IncompleteRead(0 bytes read, 1024 more expected)))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1916, in main\n    test_presigned_get_object_response_headers(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1302, in test_presigned_get_object_response_headers\n    raise Exception(err)\nException: HTTPConnectionPool(host='72.28.97.55', port=30465): Max retries exceeded with url: /minio-py-test-a169171a-ffc6-4af4-92ef-8a6aab49a0f4/dca89c2a-69bb-4dbd-a2e7-3045402f4dda?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20190719%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190719T164239Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&response-content-language=en_US&response-content-type=text%2Fplain&X-Amz-Signature=bcc6faff1fe17d1779e4f6f89ac95367a29c7ae25e43ef3b02dbf3593e90b64b (Caused by ProtocolError('Connection broken: IncompleteRead(0 bytes read, 1024 more expected)', IncompleteRead(0 bytes read, 1024 more expected)))\n",
  "args": {
    "object_name": "dca89c2a-69bb-4dbd-a2e7-3045402f4dda",
    "presigned_get_object_url": "http://72.28.97.55:30465/minio-py-test-a169171a-ffc6-4af4-92ef-8a6aab49a0f4/dca89c2a-69bb-4dbd-a2e7-3045402f4dda?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20190719%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190719T164239Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&response-content-language=en_US&response-content-type=text%2Fplain&X-Amz-Signature=bcc6faff1fe17d1779e4f6f89ac95367a29c7ae25e43ef3b02dbf3593e90b64b",
    "bucket_name": "minio-py-test-a169171a-ffc6-4af4-92ef-8a6aab49a0f4"
  },
  "message": "HTTPConnectionPool(host='72.28.97.55', port=30465): Max retries exceeded with url: /minio-py-test-a169171a-ffc6-4af4-92ef-8a6aab49a0f4/dca89c2a-69bb-4dbd-a2e7-3045402f4dda?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20190719%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190719T164239Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&response-content-language=en_US&response-content-type=text%2Fplain&X-Amz-Signature=bcc6faff1fe17d1779e4f6f89ac95367a29c7ae25e43ef3b02dbf3593e90b64b (Caused by ProtocolError('Connection broken: IncompleteRead(0 bytes read, 1024 more expected)', IncompleteRead(0 bytes read, 1024 more expected)))",
  "name": "minio-py:test_presigned_get_object_response_headers",
  "status": "FAIL",
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)"
}

Executed 11 out of 14 tests successfully.
```
<a name="ffcc0600-c2c2-408e-b444-7af757e2da99"></a>
### 7950-7434a73/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:30727
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e1bcf187cf76:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 3 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 48 seconds
(4/14) Running aws-sdk-ruby tests ... done in 13 seconds
(5/14) Running awscli tests ... done in 2 minutes and 32 seconds
(6/14) Running healthcheck tests ... done in 1 seconds
(7/14) Running mc tests ... FAILED in 2 minutes and 39 seconds
{
  "name": "mc",
  "duration": "130177",
  "function": "test_mirror_list_objects_storage_class",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-29734 --quiet --no-color mirror --storage-class REDUCED_REDUNDANCY /mint/data myminio/mc-test-bucket-3676  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-3676/datafile-1-MB`\n`/mint/data/datafile-0-b` -> `myminio/mc-test-bucket-3676/datafile-0-b`\n`/mint/data/datafile-1-kB` -> `myminio/mc-test-bucket-3676/datafile-1-kB`\n`/mint/data/datafile-1-b` -> `myminio/mc-test-bucket-3676/datafile-1-b`\n`/mint/data/datafile-10-MB` -> `myminio/mc-test-bucket-3676/datafile-10-MB`\n`/mint/data/datafile-1.03-MB` -> `myminio/mc-test-bucket-3676/datafile-1.03-MB`\n`/mint/data/datafile-10-kB` -> `myminio/mc-test-bucket-3676/datafile-10-kB`\n`/mint/data/datafile-100-kB` -> `myminio/mc-test-bucket-3676/datafile-100-kB`\n`/mint/data/datafile-129-MB` -> `myminio/mc-test-bucket-3676/datafile-129-MB`\n`/mint/data/datafile-11-MB` -> `myminio/mc-test-bucket-3676/datafile-11-MB`\n`/mint/data/datafile-5-MB` -> `myminio/mc-test-bucket-3676/datafile-5-MB`\n`/mint/data/datafile-33-kB` -> `myminio/mc-test-bucket-3676/datafile-33-kB`\n`/mint/data/datafile-6-MB` -> `myminio/mc-test-bucket-3676/datafile-6-MB`\n`/mint/data/datafile-5243880-b` -> `myminio/mc-test-bucket-3676/datafile-5243880-b`\n`/mint/data/datafile-65-MB` -> `myminio/mc-test-bucket-3676/datafile-65-MB`\nmc: <ERROR> Failed to copy `/mint/data/datafile-129-MB`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-65-MB`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-10-MB`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-11-MB`. Please reduce your request\nTotal: 233.17 MiB, Transferred: 501.68 MiB, Speed: 3.86 MiB/s"
}

Executed 6 out of 14 tests successfully.
```

--- a/docs/distributed/DESIGN.md
+++ b/docs/distributed/DESIGN.md
@@ -45,7 +45,7 @@ Expansion of ellipses and choice of erasure sets based on this expansion is an a
 
 - *If total disks has many common divisors the algorithm chooses the minimum amounts of erasure sets possible for a erasure set size of any N*.  In the example with 1024 disks - 4, 8, 16 are GCD factors. With 16 disks we get a total of 64 possible sets, with 8 disks we get a total of 128 possible sets, with 4 disks we get a total of 256 possible sets. So algorithm automatically chooses 64 sets, which is *16 * 64 = 1024* disks in total.
 
-- In this algorithm, we also make sure that we spread the disks out evenly for example. MinIO server expands ellipses passed as arguments. Here is a sample expansion to demonstrate the process.
+- In this algorithm, we also make sure that we spread the disks out evenly. MinIO server expands ellipses passed as arguments. Here is a sample expansion to demonstrate the process.
 
 ```
 minio server http://host{1...4}/export{1...8}
@@ -87,7 +87,7 @@ Expected expansion
 > http://host4/export8
 ```
 
-A noticeable trait of this expansion is that it chooses unique hosts such that the erasure code is efficient across drives across hosts.
+A noticeable trait of this expansion is that it chooses unique hosts such that the erasure code is efficient across drives and hosts.
 
 - Choosing an erasure set for the object is decided during `PutObject()`, object names are used to find the right erasure set using the following pseudo code.
 ```go
@@ -99,7 +99,7 @@ func crcHashMod(key string, cardinality int) int {
 ```
 Input for the key is the object name specified in `PutObject()`, returns a unique index. This index is one of the erasure sets where the object will reside. This function is a consistent hash for a given object name i.e for a given object name the index returned is always the same.
 
-- Write and Read quorum are required to be satisfied only across the erasure set for an object, healing is also done per object which contains the object with in the erasure set.
+- Write and Read quorum are required to be satisfied only across the erasure set for an object. Healing is also done per object within the erasure set which contains the object.
 
 - MinIO does erasure coding at the object level not at the volume level, unlike other object storage vendors. This allows applications to choose different storage class by setting `x-amz-storage-class=STANDARD/REDUCED_REDUNDANCY` for each object uploads so effectively utilizing the capacity of the cluster. Additionally these can also be enforced using IAM policies to make sure the client uploads with correct HTTP headers.
 
@@ -112,12 +112,12 @@ Standalone erasure coded configuration with 4 sets with 16 disks each, which spa
 minio server /mnt/controller{1...4}/data{1...16}
 ```
 
-Standalone erasure coded configuration with 16 sets 16 disks per set, across mnts, across controllers.
+Standalone erasure coded configuration with 16 sets, 16 disks per set, across mounts and controllers.
 ```
 minio server /mnt{1..4}/controller{1...4}/data{1...16}
 ```
 
-Distributed erasure coded configuration with 2 sets 16 disks per set across hosts.
+Distributed erasure coded configuration with 2 sets, 16 disks per set across hosts.
 ```
 minio server http://host{1...32}/disk1
 ```

--- a/docs/distributed/DESIGN.md
+++ b/docs/distributed/DESIGN.md
@@ -1,28 +1,26 @@
-# Large Bucket Support Design Guide [![Slack](https://slack.min.io/slack?type=svg)](https://slack.min.io)
-
-This document explains the design approach, advanced use cases and limits of the large bucket feature. If you're looking to get started with large bucket support, we suggest you go through the [getting started document](https://github.com/minio/minio/blob/master/docs/large-bucket/README.md) first.
+# Distributed Server Design Guide [![Slack](https://slack.min.io/slack?type=svg)](https://slack.min.io)
+This document explains the design approach, advanced use cases and limits of the MinIO distributed server.
 
 ## Command-line
-
 ```
 NAME:
-  minio server - Start object storage server.
+  minio server - start object storage server.
 
 USAGE:
   minio server [FLAGS] DIR1 [DIR2..]
   minio server [FLAGS] DIR{1...64}
 
 DIR:
-  DIR points to a directory on a filesystem. When you want to combine multiple drives
-  into a single large system, pass one directory per filesystem separated by space.
-  You may also use a `...` convention to abbreviate the directory arguments. Remote
-  directories in a distributed setup are encoded as HTTP(s) URIs.
+  DIR points to a directory on a filesystem. When you want to combine
+  multiple drives into a single large system, pass one directory per
+  filesystem separated by space. You may also use a '...' convention
+  to abbreviate the directory arguments. Remote directories in a
+  distributed setup are encoded as HTTP(s) URIs.
 ```
 
 ## Common usage
 
 Standalone erasure coded configuration with 4 sets with 16 disks each.
-
 ```
 minio server dir{1...64}
 ```
@@ -33,38 +31,21 @@ Distributed erasure coded configuration with 64 sets with 16 disks each.
 minio server http://host{1...16}/export{1...64}
 ```
 
-## Other usages
+## Architecture
 
-### Advanced use cases with multiple ellipses
+Expansion of ellipses and choice of erasure sets based on this expansion is an automated process in MinIO. Here are some of the details of our underlying erasure coding behavior.
 
-Standalone erasure coded configuration with 4 sets with 16 disks each, which spawns disks across controllers.
-```
-minio server /mnt/controller{1...4}/data{1...16}
-```
+- Erasure coding used by MinIO is [Reed-Solomon](https://github.com/klauspost/reedsolomon) erasure coding scheme, which has a total shard maximum of 256 i.e 128 data and 128 parity. MinIO design goes beyond this limitation by doing some practical architecture choices.
 
-Standalone erasure coded configuration with 16 sets 16 disks per set, across mnts, across controllers.
-```
-minio server /mnt{1..4}/controller{1...4}/data{1...16}
-```
+- Erasure set is a single erasure coding unit within a MinIO deployment. An object is sharded within an erasure set. Erasure set size is automatically calculated based on the number of disks. MinIO supports unlimited number of disks but each erasure set can be upto 16 disks and a minimum of 4 disks.
 
-Distributed erasure coded configuration with 2 sets 16 disks per set across hosts.
-```
-minio server http://host{1...32}/disk1
-```
+- We limited the number of drives to 16 for erasure set because, erasure code shards more than 16 can become chatty and do not have any performance advantages. Additionally since 16 drive erasure set gives you tolerance of 8 disks per object by default which is plenty in any practical scenario.
 
-Distributed erasure coded configuration with rack level redundancy 32 sets in total, 16 disks per set.
-```
-minio server http://rack{1...4}-host{1...8}.example.net/export{1...16}
-```
+- Choice of erasure set size is automatic based on the number of disks available, let's say for example if there are 32 servers and 32 disks which is a total of 1024 disks. In this scenario 16 becomes the erasure set size. This is decided based on the greatest common divisor (GCD) of acceptable erasure set sizes ranging from *4, 6, 8, 10, 12, 14, 16*.
 
-Distributed erasure coded configuration with no rack level redundancy but redundancy with in the rack we split the arguments, 32 sets in total, 16 disks per set.
-```
-minio server http://rack1-host{1...8}.example.net/export{1...16} http://rack2-host{1...8}.example.net/export{1...16} http://rack3-host{1...8}.example.net/export{1...16} http://rack4-host{1...8}.example.net/export{1...16}
-```
+- *If total disks has many common divisors the algorithm chooses the minimum amounts of erasure sets possible for a erasure set size of any N*.  In the example with 1024 disks - 4, 8, 16 are GCD factors. With 16 disks we get a total of 64 possible sets, with 8 disks we get a total of 128 possible sets, with 4 disks we get a total of 256 possible sets. So algorithm automatically chooses 64 sets, which is *16 * 64 = 1024* disks in total.
 
-### Expected expansion for double ellipses
-
-MinIO server internally expands ellipses passed as arguments. Here is a sample expansion to demonstrate the process
+- In this algorithm, we also make sure that we spread the disks out evenly for example. MinIO server expands ellipses passed as arguments. Here is a sample expansion to demonstrate the process.
 
 ```
 minio server http://host{1...4}/export{1...8}
@@ -106,6 +87,50 @@ Expected expansion
 > http://host4/export8
 ```
 
+A noticeable trait of this expansion is that it chooses unique hosts such that the erasure code is efficient across drives across hosts.
+
+- Choosing an erasure set for the object is decided during `PutObject()`, object names are used to find the right erasure set using the following pseudo code.
+```go
+// hashes the key returning an integer.
+func crcHashMod(key string, cardinality int) int {
+        keyCrc := crc32.Checksum([]byte(key), crc32.IEEETable)
+        return int(keyCrc % uint32(cardinality))
+}
+```
+Input for the key is the object name specified in `PutObject()`, returns a unique index. This index is one of the erasure sets where the object will reside. This function is a consistent hash for a given object name i.e for a given object name the index returned is always the same.
+
+- Write and Read quorum are required to be satisfied only across the erasure set for an object, healing is also done per object which contains the object with in the erasure set.
+
+- MinIO does erasure coding at the object level not at the volume level, unlike other object storage vendors. This allows applications to choose different storage class by setting `x-amz-storage-class=STANDARD/REDUCED_REDUNDANCY` for each object uploads so effectively utilizing the capacity of the cluster. Additionally these can also be enforced using IAM policies to make sure the client uploads with correct HTTP headers.
+
+## Other usages
+
+### Advanced use cases with multiple ellipses
+
+Standalone erasure coded configuration with 4 sets with 16 disks each, which spawns disks across controllers.
+```
+minio server /mnt/controller{1...4}/data{1...16}
+```
+
+Standalone erasure coded configuration with 16 sets 16 disks per set, across mnts, across controllers.
+```
+minio server /mnt{1..4}/controller{1...4}/data{1...16}
+```
+
+Distributed erasure coded configuration with 2 sets 16 disks per set across hosts.
+```
+minio server http://host{1...32}/disk1
+```
+
+Distributed erasure coded configuration with rack level redundancy 32 sets in total, 16 disks per set.
+```
+minio server http://rack{1...4}-host{1...8}.example.net/export{1...16}
+```
+
+Distributed erasure coded configuration with no rack level redundancy but redundancy with in the rack we split the arguments, 32 sets in total, 16 disks per set.
+```
+minio server http://rack1-host{1...8}.example.net/export{1...16} http://rack2-host{1...8}.example.net/export{1...16} http://rack3-host{1...8}.example.net/export{1...16} http://rack4-host{1...8}.example.net/export{1...16}
+```
 ## Backend `format.json` changes
 
 `format.json` has new fields
@@ -186,5 +211,5 @@ type formatXLV2 struct {
 
 ## Limits
 
-- Minimum of 4 disks are needed for erasure coded configuration.
+- Minimum of 4 disks are needed for any erasure coded configuration.
 - Maximum of 32 distinct nodes are supported in distributed configuration.

--- a/docs/distributed/README.md
+++ b/docs/distributed/README.md
@@ -67,7 +67,6 @@ __NOTE:__ `{1...n}` shown have 3 dots! Using only 2 dots `{1..32}` will be inter
 To test this setup, access the MinIO server via browser or [`mc`](https://docs.min.io/docs/minio-client-quickstart-guide).
 
 ## Explore Further
-- [MinIO Large Bucket Support Guide](https://docs.min.io/docs/minio-large-bucket-support-quickstart-guide)
 - [MinIO Erasure Code QuickStart Guide](https://docs.min.io/docs/minio-erasure-code-quickstart-guide)
 - [Use `mc` with MinIO Server](https://docs.min.io/docs/minio-client-quickstart-guide)
 - [Use `aws-cli` with MinIO Server](https://docs.min.io/docs/aws-cli-with-minio)

